activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb

```diff
@@ -634,7 +634,11 @@ def use_insert_returning?
       # Returns the version of the connected PostgreSQL server.
       def get_database_version # :nodoc:
         with_raw_connection do |conn|
-          conn.server_version
+          version = conn.server_version
+          if version == 0
+            raise ActiveRecord::ConnectionFailed, "Could not determine PostgreSQL version"
```

Comment:
I wasn't sure what the appropriate error here was. I went with this over `ActiveRecord::ConnectionNotEstablished` only because of this snippet of code indicating the former is used for libpq related errors: https://github.com/rails/rails/blob/049d1af38d19968fc8e8648989b0fb8aae2caa71/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L804-L814

Note that libpq may returr 0 either of there is no connection or if an existing connection is bad: https://github.com/postgres/postgres/blob/7f24c0274385ea3d4c797cc2fd60d9a55f3a8d74/src/interfaces/libpq/fe-connect.c#L7542-L7550